design/proposed/2020-openstackserver-controller.md

```diff
@@ -0,0 +1,145 @@
+
+# Potential solutions
+
+## Do nothing
+
+The current solution has sufficed in practice for quite a long time. Additionally, we now have much better test coverage of bastion operations than we have had in the past, so we should hopefully catch issues sooner.
+
+The primary issue with this is the drag it places on bugfixes and new development, because the code flow of bastion deletion in particular can be hard to reason about.
+
+## Create an OpenStackMachine for the bastion
+
+This is an attractive solution because it better models the independent lifecycle of the bastion. When a new bastion is created we create a new OpenStackMachine. If the bastion spec is updated we delete the old one and create a new one, with the old and new objects having their own, independent state.
+
+The problem with this approach is that OpenStackMachine has other contracts. It cannot be created independently, but requires a CAPI Machine object. Other controllers have expectations of Machine objects, for example that they will eventually have a NodeRef.
+
+Creating a Machine/OpenStackMachine object for the bastion could negatively impact other CAPI users.
+
+## Create a new controller for the bastion
+
+This is essentially the same as using OpenStackMachine, except using a new CRD which can be independent of other CAPI objects.
+
+The obvious disadvantage is that it requires a new CRD and controller.
+
+# Proposal: An OpenStackServer CRD and controller
+
+## Philosophy
+
+We will create a new controller which is capable, in principal, of doing all of the OpenStack resource creation tasks common to bastion and OpenStackMachine creation. That is:
+
+* Resolve OpenStack resource parameters to specific resources (e.g. image, networks, subnets, security groups, server groups)
+* Create and delete ports
+* Create and delete volumes
+* Create and delete the server
+
+It will have no dependency on Machine or OpenStackCluster. So things it will explicitly not do include:
+
+* Attaching the Bastion floating IP
+* Creating a loadbalancer member
+* Adding a port for the default cluster network if no ports are provided
+* Referencing an AvailabilityZone from the Machine object
+* Adding default tags from the OpenStackCluster object
+  
+## API
+
+Based on OpenStackMachineSpec, with modifications to accomodate the above.
+
+
+```go
+type OpenStackServerSpec struct {
```

Comment:
is there anything we need to do for existing customer to move to here? transparent or need define those ?

Reply:
That's a great point. I should make it clear in the doc that this change will happen automatically and the end user does not need to take any action. I'll update the doc.